stores/walmart.py

```diff
@@ -0,0 +1,390 @@
+                    self.password + Keys.RETURN
+                )
+                time.sleep(2)
+            except:
+                pass
+
+        if self.driver.find_elements_by_xpath('//*[@id="global-error"]'):
+            log.error("Login failed, check your username in walmart_config.json")
+            time.sleep(240)
+            exit(1)
+        log.info(f"Logged in as {self.username}")
+
+    def run_item(self, delay=5, test=False):
+        checkout_success = False
+        while not checkout_success:
+            pop_list = []
+            for i in range(len(self.url_list)):
+                for url in self.url_list[i]:
+                    self.check_captcha(url)
+                    checkout_success = self.check_stock(url, self.reserve[i])
+                    if checkout_success:
+                        log.info(f"Attempting to buy {url}")
+                        if self.checkout(test=test):
+                            log.info(f"bought {url}")
+                            pop_list.append(url)
+                            break
+                        else:
+                            log.info(f"checkout for {url} failed")
+                            checkout_success = False
+                    time.sleep(1)
+            if pop_list:
+                for url in pop_list:
+                    for i in range(len(self.url_list)):
+                        if url in self.url_list[i]:
+                            self.url_list.pop(i)
+                            self.reserve.pop(i)
+                            break
+            if self.url_list:  # keep bot going if additional URLs left
+                checkout_success = False
+                # log.info("Additional lists remaining, bot will continue")
+
+    def check_captcha(self, url):
+        if self.driver.title in CAPTCHA_URL:
+            self.defeat_captcha(self.driver.current_url)
+
+    def check_stock(self, url, reserve):
+        log.info("Checking stock for items.")
+        f = furl(WALMART_URLS["OFFER_URL"] + url)
+        self.driver.get(f.url)
+        self.check_captcha(url)
+        try:
+            element = self.check_exists_by_xpath(
+                "/html/body/div[1]/div[1]/div/div[2]/div/div[1]/div[1]/div[1]/div/div/div/div/div[3]/div[5]/div/div[3]/div/div[2]/div[2]/div[1]/section/div[1]/div[3]/button/span"
+            )
+            str_price = self.driver.find_element_by_xpath('//*[@id="price"]')
+            log.info("Found price " + str_price.text)
+        except NoSuchElementException:
+            return False
+
+        if not element:
+            log.info("Product not available")
+            self.check_stock(url, reserve)
+
+        price = parse_price(str_price.text)
+        priceFloat = price.amount
+        if priceFloat is None:
+            log.error("Error reading price information on row.")
+        elif priceFloat <= reserve:
+            if element:
```

Comment:
```
if priceFloat is None:
    log.error("Error reading price information on row.")
```
It could probably be removed but I left it in as a fail-safe. The try section just above doesn't check if price exists. It only tries to set str_price to whatever is at the XPath. The code checks for the existence of the "Add to Cart" button:

`element = self.check_exists_by_xpath(
                "/html/body/div[1]/div[1]/div/div[2]/div/div[1]/div[1]/div[1]/div/div/div/div/div[3]/div[5]/div/div[3]/div/div[2]/div[2]   /div[1]/section/div[1]/div[3]/button/span")`

Better code might be:
```
try:
    addtocart = self.check_exists_by_xpath(
        "/html/body/div[1]/div[1]/div/div[2]/div/div[1]/div[1]/div[1]/div/div/div/div/div[3]/div[5]/div/div[3]/div/div[2]/div[2]/div[1]     /section/>")
    if self.driver.find_element_by_xpath('//*[@id="price"]'):
        str_price = self.driver.find_element_by_xpath('//*[@id="price"]')
        log.info("Found price " + str_price.text)
    else:
        log.error("Error reading price information on row.")
except NoSuchElementException:
    return False

if not addtocart:
    log.info("Product not available")
    self.check_stock(url, reserve)

price = parse_price(str_price.text)
priceFloat = price.amount

if priceFloat <= reserve:
    if addtocart:
        log.info("Item in stock and under reserve!")
        log.info("Clicking add to cart")
        addtocart.click()
    return True   
return False
```

Reply:
it looks like github pulls in the previous 3 lines on the comment - I think the comment is typically related to the last line (unless more than 4 lines are included?). My comment was specifically related to line 233 where you check ```if element:``` but we can't get to that point unless we pass the ```if not element:``` check on line 224.

In any case, I'd probably do something like this:
```
def check_stock(self, url, reserve):
    log.info("Checking stock for items.")
    f = furl(WALMART_URLS["OFFER_URL"] + url)
    self.driver.get(f.url)
    self.check_captcha(url)  # need something to verify we actually passed captcha, right?
    element = self.check_exists_by_xpath( # I'm not sure why we need to do this vs. find_element_by_xpath, but I'm going with what you had
        "/html/body/div[1]/div[1]/div/div[2]/div/div[1]/div[1]/div[1]/div/div/div/div/div[3]/div[5]/div/div[3]/div/div[2]/div[2]/div[1]/section/div[1]/div[3]/button/span"
    ) # I pulled this out of try because you were already testing for exception in check_exists_by_xpath
    if element:
        try:
            str_price = self.driver.find_element_by_xpath('//*[@id="price"]')
            log.info("Found price " + str_price.text)
        except NoSuchElementException:
            log.error("Error reading price information on row")
            return False

        price = parse_price(str_price.text)
        priceFloat = price.amount
        if priceFloat is None:
            log.error("Error reading price information on row.")
        elif priceFloat <= reserve:
            log.info("Item in stock and under reserve!")
            log.info("Clicking add to cart")
            element.click()
            return True
    return False
```